File: android/MpOS API/src/br/ufc/mdcc/mpos/net/rpc/RpcClient.java
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package br.ufc.mdcc.mpos.net.rpc;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.OutputStream;
import java.lang.reflect.Method;
import java.net.ConnectException;

import br.ufc.mdcc.mpos.net.Protocol;
import br.ufc.mdcc.mpos.net.core.ClientTcp;
import br.ufc.mdcc.mpos.net.core.FactoryClient;
import br.ufc.mdcc.mpos.net.endpoint.ServerContent;
import br.ufc.mdcc.mpos.net.exceptions.MissedEventException;
import br.ufc.mdcc.mpos.net.rpc.model.Code;
import br.ufc.mdcc.mpos.net.rpc.model.RpcProfile;
import br.ufc.mdcc.mpos.net.rpc.util.DebugBufferedInputStream;
import br.ufc.mdcc.mpos.net.rpc.util.RpcException;
import br.ufc.mdcc.mpos.net.rpc.util.RpcSerializable;

/**
 * This is a client implementation used for rpc call, under 
 * TCP protocol.
 * 
 * @author Philipp B. Costa
 */
public final class RpcClient {
	private final int BUFFER_SIZE = 4096;
	private final RpcProfile profile;
	private final byte dataFlag[] = new byte[1];

	private ServerContent server;

	private InputStream inputStream;
	private OutputStream outputStream;

	public RpcClient() {
		this.profile = new RpcProfile();
		this.server = null;
	}

	public void setupServer(ServerContent server) {
		this.server = server;
	}

	public Object call(Object objOriginal, Method method, Object params[]) throws RpcException, ConnectException {
		return call(false, objOriginal, method.getName(), params);
	}

	public Object call(boolean manualSerialization, Object objOriginal, String methodName, Object params[]) throws RpcException, ConnectException {
		return call(false, manualSerialization, objOriginal, methodName, params);
	}

	public Object call(boolean needProfile, boolean manualSerialization, Object objOriginal, String methodName, Object params[]) throws RpcException, ConnectException {
		if (server == null) {
			throw new ConnectException("Need to setup any server for use the RPC Client");
		}

		ClientTcp client = (ClientTcp) FactoryClient.getInstance(Protocol.TCP_STREAM);

		try {
			client.connect(server.getIp(), server.getRpcServicePort());

			sent(client.getOutputStream(), manualSerialization, needProfile, objOriginal, methodName, params);
			ResponseRemotable response = receive(client.getInputStream(), objOriginal, methodName);

			if (response.code == Code.OK) {
				close(client);// reduce server wait client finish!
				return response.methodReturn;
			} else if (response.code == Code.METHOD_THROW_ERROR) {
				close(client);
				throw new RpcException("[Server]: Remote method thrown some errors\n" + response.except);
			} else {
				close(client);
				throw new RpcException("[Server]: RPC call terminated with some errors!\n" + response.except);
			}
		} catch (ConnectException e) {
			close(client);
			throw new ConnectException("Failed to connect to server: " + server.getIp() + ":" + server.getRpcServicePort());
		} catch (IOException e) {
			close(client);
			throw new RpcException("RPC suffer with some I/O error, check yours dependences: " + objOriginal.getClass().getName());
		} catch (MissedEventException e) {
			close(client);
			throw new RpcException("ERROR: TCP thread was started?");
		} catch (ClassNotFoundException e) {
			close(client);
			throw new RpcException("ERROR: Deploy jar wasn't updated from assets!");
		}
	}

	private void sent(OutputStream os, boolean manualSerialization, boolean debug, Object objOriginal, String methodName, Object params[]) throws IOException {
		if (manualSerialization) {
			dataFlag[0] = debug ? Code.DATASTREAMDEBUG : Code.DATASTREAM;
			os.write(dataFlag);

			DataOutputStream dos = new DataOutputStream(new BufferedOutputStream(os, BUFFER_SIZE));
			outputStream = dos;

			dos.writeUTF(objOriginal.getClass().getName());
			dos.writeUTF(methodName);
			((RpcSerializable) objOriginal).writeMethodParams(dos, methodName, params);
			dos.flush();
		} else {
			dataFlag[0] = debug ? Code.OBJECTSTREAMDEBUG : Code.OBJECTSTREAM;
			os.write(dataFlag);

			ObjectOutputStream oos = new ObjectOutputStream(new BufferedOutputStream(os, BUFFER_SIZE));
			outputStream = oos;

			oos.writeUTF(objOriginal.getClass().getName());
			oos.writeUTF(methodName);
			oos.writeObject(params);

			oos.flush();
		}
	}

	private ResponseRemotable receive(InputStream is, Object objOriginal, String methodName) throws IOException, ClassNotFoundException {
		ResponseRemotable response = new ResponseRemotable();

		is.read(dataFlag);
		if (dataFlag[0] == Code.DATASTREAM) {
			DataInputStream dis = new DataInputStream(new BufferedInputStream(is, BUFFER_SIZE));
			inputStream = dis;

			int code = dis.readInt();
			response.code = code;
			if (code == Code.OK) {
				response.methodReturn = ((RpcSerializable) objOriginal).readMethodReturn(dis, methodName);
			} else if (code == Code.METHOD_THROW_ERROR) {
				response.except = dis.readUTF();
			} else {
				throw new IOException("Code different from expected: " + code);
			}
		} else if (dataFlag[0] == Code.OBJECTSTREAM) {
			ObjectInputStream ois = new ObjectInputStream(new BufferedInputStream(is, BUFFER_SIZE));
			inputStream = ois;

			response.code = ois.readInt();
			response.methodReturn = ois.readObject();
		} else if (dataFlag[0] == Code.DATASTREAMDEBUG) {
			DebugBufferedInputStream bufInput = new DebugBufferedInputStream(is, BUFFER_SIZE);
			DataInputStream dis = new DataInputStream(bufInput);
			inputStream = dis;

			int code = dis.readInt();
			response.code = code;
			if (code == Code.OK) {
				profile.setUploadSize(dis.readInt());
				profile.setUploadTime(dis.readLong());
				profile.setExecutionCpuTime(dis.readLong());
				long initDownloadTime = System.currentTimeMillis();
				response.methodReturn = ((RpcSerializable) objOriginal).readMethodReturn(dis, methodName);
				profile.setDonwloadTime(System.currentTimeMillis() - initDownloadTime);
				profile.setDownloadSize(bufInput.getTotalReadData());
			} else if (code == Code.METHOD_THROW_ERROR) {
				response.except = dis.readUTF();
			} else {
				response.except = "Code different from expected: " + code;
			}
		} else if (dataFlag[0] == Code.OBJECTSTREAMDEBUG) {
			DebugBufferedInputStream bufInput = new DebugBufferedInputStream(is, BUFFER_SIZE);
			ObjectInputStream ois = new ObjectInputStream(bufInput);
			inputStream = ois;

			response.code = ois.readInt();
			profile.setUploadSize(ois.readInt());
			profile.setUploadTime(ois.readLong());
			profile.setExecutionCpuTime(ois.readLong());

			long initDownloadTime = System.currentTimeMillis();
			response.methodReturn = ois.readObject();
			profile.setDonwloadTime(System.currentTimeMillis() - initDownloadTime);
			profile.setDownloadSize(bufInput.getTotalReadData());
		}

		return response;
	}

	private void close(ClientTcp client) {
		try {
			outputStream.close();
			inputStream.close();
			client.close();
		} catch (IOException e) {
		}
	}

	public RpcProfile getProfile() {
		return profile;
	}
	
	//only internal use!
	private final class ResponseRemotable {
		int code;
		Object methodReturn;
		String except;
	}
}

File: windows_phone/MpOS API/MpOS/Net/Rpc/RpcClient.cs
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Reflection;
using Ufc.MpOS.Net.Core;
using Ufc.MpOS.Net.Endpoint;
using Ufc.MpOS.Net.Rpc.Model;
using Ufc.MpOS.Net.Rpc.Util;
using Ufc.MpOS.Util.Exceptions;

namespace Ufc.MpOS.Net.Rpc
{
	sealed class RpcClient
	{
		private ServerContent server;
		private byte dataFlag = 0;

		private RpcProfile profile;
		public RpcProfile Profile
		{
			get
			{
				return profile;
			}
		}

		public RpcClient()
		{
			profile = new RpcProfile();
			this.server = null;
		}

		public void SetupServer(ServerContent server)
		{
			this.server = server;
		}

		public object Call(Object objOriginal, MethodInfo method, object[] methodParams)
		{
			return Call(false, objOriginal, method.Name, methodParams);
		}

		public object Call(bool manualSerialization, object objOriginal, string methodName, object[] methodParams)
		{
			return Call(false, manualSerialization, objOriginal, methodName, methodParams);
		}

		public object Call(bool needProfile, bool manualSerialization, Object objOriginal, String methodName, object[] methodParams)
		{
			ClientTcp socket = null;

			if (server == null)
			{
				throw new NetworkException("Need to setup any server for use the RPC Client");
			}

			try
			{
				socket = new ClientTcp();
				socket.Connect(server.Ip, server.RpcServicePort);

				Sent(socket, manualSerialization, needProfile, objOriginal, methodName, methodParams);
				ResponseRemotable response = Receive(socket, objOriginal, methodName);

				if (response.Code == Code.OK)
				{
					return response.MethodReturn;
				}
				else if (response.Code == Code.METHOD_THROW_ERROR)
				{
					throw new RpcException("[Server]: Remote method thrown some errors\n" + response.Except);
				}
				else
				{
					throw new RpcException("[Server]: RPC call terminated with some errors!\n" + response.Except);
				}
			}
			catch (SocketException e)
			{
				throw new ConnectException("Failed to connect to server: " + server.Ip + ":" + server.RpcServicePort, e);
			}
			finally
			{
				socket.Close();
			}
		}

		private void Sent(ClientTcp socket, bool manualSerialization, bool debug, object objOriginal, string methodName, object[] methodParams)
		{
			using (MemoryStream memoryStream = new MemoryStream(1024))
			{
				using (BinaryWriter writer = new BinaryWriter(memoryStream))
				{
					if (manualSerialization)
					{
						dataFlag = debug ? Code.CUSTOMSTREAMDEBUG : Code.CUSTOMSTREAM;

						writer.Write(objOriginal.GetType().FullName);
						writer.Write(methodName);
						((RpcSerializable)objOriginal).WriteMethodParams(writer, methodName, methodParams);
					}
					else
					{
						dataFlag = debug ? Code.BSONSTREAMDEBUG : Code.BSONSTREAM;

						writer.Write(objOriginal.GetType().FullName);
						writer.Write(methodName);

						byte[] bsonData = BsonFormatter.Serialize(methodParams);
						writer.Write(bsonData.Length);
						writer.Write(bsonData);
					}

					writer.Flush();

					byte[] headerSize = new byte[5];
					headerSize[4] = dataFlag;
					Array.Copy(BitConverter.GetBytes(memoryStream.Length), headerSize, 4);


					byte[] sentStream = new byte[(int)memoryStream.Length + 5];
					headerSize.CopyTo(sentStream, 0);
					memoryStream.ToArray().CopyTo(sentStream, headerSize.Length);

					socket.Sent(sentStream);
				}
			}
			//Debug.WriteLine(">> RpcClient Sent Data");
		}

		//async method can't receive out params!
		private ResponseRemotable Receive(ClientTcp socket, object objOriginal, string methodName)
		{
			//Debug.WriteLine(">> start RpcClient Receive Data");

			ResponseRemotable response = new ResponseRemotable();
			BinaryReader reader = null;

			long downloadTime = 0L;
			int downloadSize = 0;
			if (dataFlag == Code.CUSTOMSTREAM || dataFlag == Code.BSONSTREAM)
			{
				reader = new BinaryReader(socket.ReceiveWithHeaderSize());
			}
			else
			{
				reader = new BinaryReader(socket.ReceiveWithHeaderDebug(ref downloadTime, ref downloadSize));
			}

			//Debug.WriteLine(">> RpcClient Receive wait first flag");
			dataFlag = reader.ReadByte();
			//Debug.WriteLine(">> RpcClient Receive end first flag");

			if (dataFlag == Code.CUSTOMSTREAM)
			{
				response.Code = Code.OK;
				response.MethodReturn = ((RpcSerializable)objOriginal).ReadMethodReturn(reader, methodName);
			}
			else if (dataFlag == Code.BSONSTREAM)
			{
				response.Code = Code.OK;
				response.MethodReturn = GetMethodReturn(reader);
			}
			else if (dataFlag == Code.CUSTOMSTREAMDEBUG)
			{
				response.Code = Code.OK;

				profile.UploadSize = reader.ReadInt32();
				profile.UploadTime = reader.ReadInt64();
				profile.ExecutionCpuTime = reader.ReadInt64();
				response.MethodReturn = ((RpcSerializable)objOriginal).ReadMethodReturn(reader, methodName);

				profile.DonwloadTime = downloadTime;
				profile.DownloadSize = downloadSize;
			}
			else if (dataFlag == Code.BSONSTREAMDEBUG)
			{
				response.Code = Code.OK;

				profile.UploadSize = reader.ReadInt32();
				profile.UploadTime = reader.ReadInt64();
				profile.ExecutionCpuTime = reader.ReadInt64();
				response.MethodReturn = GetMethodReturn(reader);

				profile.DonwloadTime = downloadTime;
				profile.DownloadSize = downloadSize;
			}
			else if (dataFlag == Code.METHOD_THROW_ERROR)
			{
				response.Code = dataFlag;
				response.Except = reader.ReadString();
			}
			else
			{
				response.Code = Code.SERVER_ERROR;
				response.Except = "Code different from expected: " + dataFlag;
			}

			//Debug.WriteLine(">> end RpcClient Receive Data");

			reader.Dispose();
			return response;
		}

		private object GetMethodReturn(BinaryReader streamReader)
		{
			int bsonLength = streamReader.ReadInt32();
			byte[] bsonData = new byte[bsonLength];
			streamReader.Read(bsonData, 0, bsonLength);
			return BsonFormatter.Deserialize(bsonData);
		}

		//only internal use!
		private sealed class ResponseRemotable
		{
			public byte Code { get; set; }
			public object MethodReturn { get; set; }
			public string Except { get; set; }
		}
	}
}